src/plugins/data/public/query/state_sync/use_sync_state_with_url.test.ts

```diff
@@ -0,0 +1,122 @@
+
+jest.mock('./sync_state_with_url');
+
+const setupMock = coreMock.createSetup();
+const startMock = coreMock.createStart();
+
+setupMock.uiSettings.get.mockImplementation((key: string) => {
+  switch (key) {
+    case 'defaultIndex':
+      return 'logstash-*';
+    case UI_SETTINGS.FILTERS_PINNED_BY_DEFAULT:
+      return true;
+    case 'timepicker:timeDefaults':
+      return { from: 'now-15m', to: 'now' };
+    case 'search:queryLanguage':
+      return 'kuery';
+    case UI_SETTINGS.TIMEPICKER_REFRESH_INTERVAL_DEFAULTS:
+      return { pause: false, value: 0 };
+    case UI_SETTINGS.QUERY_ENHANCEMENTS_ENABLED:
+      return false;
+    case UI_SETTINGS.SEARCH_MAX_RECENT_DATASETS:
+      return 4;
+    default:
+      throw new Error(`sync_query test: not mocked uiSetting: ${key}`);
+  }
+});
+
+describe('use_sync_query_state_with_url', () => {
+  let queryServiceStart: QueryStart;
+  let filterManager: FilterManager;
+  let timefilter: TimefilterContract;
+  let osdUrlStateStorage: IOsdUrlStateStorage;
+  let history: History;
+  let indexPatternsMock: IndexPatternsService;
+
+  beforeEach(() => {
+    indexPatternsMock = ({
+      get: jest.fn(),
+    } as unknown) as IndexPatternsService;
+  });
+
+  let filterManagerChangeSub: Subscription;
+  let filterManagerChangeTriggered = jest.fn();
+  let mockSearchInterceptor: jest.Mocked<ISearchInterceptor>;
+
+  let gF: Filter;
+  let aF: Filter;
+
+  beforeEach(() => {
+    const queryService = new QueryService();
+    queryService.setup({
+      uiSettings: setupMock.uiSettings,
+      storage: new DataStorage(window.localStorage, 'opensearch_dashboards.'),
+      sessionStorage: new DataStorage(window.sessionStorage, 'opensearch_dashboards.'),
+      defaultSearchInterceptor: mockSearchInterceptor,
+      application: setupMock.application,
+      notifications: setupMock.notifications,
+    });
+    queryServiceStart = queryService.start({
+      indexPatterns: indexPatternsMock,
+      uiSettings: startMock.uiSettings,
+      storage: new DataStorage(window.localStorage, 'opensearch_dashboards.'),
+      savedObjectsClient: startMock.savedObjects.client,
+      application: startMock.application,
+      notifications: startMock.notifications,
+    });
+    filterManager = queryServiceStart.filterManager;
+    timefilter = queryServiceStart.timefilter.timefilter;
+
+    filterManagerChangeTriggered = jest.fn();
+    filterManagerChangeSub = filterManager.getUpdates$().subscribe(filterManagerChangeTriggered);
+
+    window.location.href = '/';
+    history = createBrowserHistory();
+    osdUrlStateStorage = createOsdUrlStateStorage({ useHash: false, history });
+
+    gF = getFilter(FilterStateStore.GLOBAL_STATE, true, true, 'key1', 'value1');
+    aF = getFilter(FilterStateStore.APP_STATE, true, true, 'key3', 'value3');
+  });
+  afterEach(() => {
+    filterManagerChangeSub.unsubscribe();
+  });
+
+  it('Should invoke connectStorageToQueryState', () => {
```

Comment:
I like the solution, but since the main goal of the hook is to clear up the subscription, shouldn't there also be a test for that?